fetch.bs

```diff
@@ -1781,6 +1781,12 @@ which is "<code>omit</code>", "<code>same-origin</code>", or
 <dfn export for=request id=concept-request-use-url-credentials-flag>use-URL-credentials flag</dfn>.
 Unless stated otherwise, it is unset.
 
+<p class=note>This flag controls whether the <a for=/>request</a>'s <a for=request>URL</a>'s
+<a for=url>username</a> and <a for=url>password</a> will be used to look up an
+<a>authentication entry</a> or not. Modern specifications avoid setting it, since putting
+credentials in <a for=/>URLs</a> is discouraged, but some older features set it for compatibility
+reasons.
```

Comment:
This is incorrect. If this flag is true and URL contains username/password, we'll use the URL's username/password instead of the authentication entry (which is stored in a side table).

Reply:
Eek, OK, I think I understand the relevant conditionals now. So let me break it down to make sure.

There are three input conditions which interact with the UUC flag:

1. authentication entry exists in the side table
2. URL includes credentials
3. isAuthenticationFetch is true, i.e., are we doing a second fetch after a 401 with WWW-Authenticate

If the UUC flag is not set, then:

- (1) + (2) + (3) => use side table
- (1) + (2) => use side table
- (1) + (3) => use side table
- (2) + (3) => use credentials from the URL
- (1) => use side table
- (2) => use credentials from the URL
- (3) => no credentials

If the UUC flag is set, then:

- (1) + (2) + (3) => **use credentials from the URL**
- (1) + (2) => **no credentials**
- (1) + (3) => use side table
- (2) + (3) => use credentials from the URL
- (1) => use side table
- (2) => no credentials
- (3) => no credentials

Does this seem right? The (1) + (2) case is particularly a bit surprising of a change.

Reply:
Yeah, this seems to match the spec, but I'm not sure it's what we want or whether it matches browsers.

If you mean "(1) + (2) => no credentials" from the second list, that seems somewhat reasonable to me. Presumably that would then result in a 401 after which the URL credentials would be used, no?

I think what's more surprising is that we end up using URL credentials when the UUC flag is not set. That seems like a bug of sorts. (Some user agents associate credentials with a request by serializing them in the URL, which is perhaps where some of this stems from. It's been a long time.)

Reply:
> Presumably that would then result in a 401 after which the URL credentials would be used, no?

Eh, seems like it depends on whether the 401 results in a re-request with the prompted credentials in the URL, or in the side table. Which I'm not sure about.

> I think what's more surprising is that we end up using URL credentials when the UUC flag is not set. That seems like a bug of sorts.

My interpretation (and what I'm revising this PR to do) is that the current spec uses the flag to preference URL credentials over side table credentials, if both are present. If only one of the two is present, then it still gets used, regardless of the flag.

But, that's not how Chromium seems to behave per #1496 at least, so yeah, maybe a bug. I'll comment over there. We could add an XXX pointing to that issue if you think it's a good idea?